File: fedorova-t.v-kt-41-22/Database/Configurations/TeacherConfiguration.vb
Imports fedorova_t.v_kt_41_22.Models
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Metadata.Builders



Namespace fedorova_t.v_kt_41_22.Database.Configurations
    Public Class TeacherConfiguration
        Implements IEntityTypeConfiguration(Of Teacher)
        Public Sub Configure(builder As EntityTypeBuilder(Of Teacher)) Implements IEntityTypeConfiguration(Of Teacher).Configure
            builder.ToTable("Teachers")
            builder.HasKey(Function(t) t.Id)

            builder.Property(Function(t) t.FirstName).HasMaxLength(50).IsRequired()
            builder.Property(Function(t) t.LastName).HasMaxLength(50).IsRequired()

            builder.HasOne(Of AcademicDegree)(Function(t) CType(t.AcademicDegree, AcademicDegree?)).WithMany(Function(d) d.Teachers).HasForeignKey(CType(Function(t) CObj(t.AcademicDegreeId), Expressions.Expression(Of Func(Of Teacher, Object?)))).OnDelete(DeleteBehavior.Restrict).IsRequired(False)

            builder.HasOne(Of Position)(Function(t) CType(t.Position, Position?)).WithMany(Function(p) p.Teachers).HasForeignKey(CType(Function(t) CObj(t.PositionId), Expressions.Expression(Of Func(Of Teacher, Object?)))).OnDelete(DeleteBehavior.Restrict).IsRequired(False)

            builder.HasOne(Of Department)(Function(t) CType(t.Department, Department?)).WithMany(Function(d) d.Teachers).HasForeignKey(CType(Function(t) CObj(t.DepartmentId), Expressions.Expression(Of Func(Of Teacher, Object?)))).OnDelete(DeleteBehavior.Cascade)
        End Sub

    End Class
End Namespace

File: fedorova-t.v-kt-41-22/Database/Configurations/TeacherConfiguration.cs
using fedorova_t.v_kt_41_22.Models;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
using fedorova_t.v_kt_41_22.Database.Helpers;


namespace fedorova_t.v_kt_41_22.Database.Configurations
{
    public class TeacherConfiguration : IEntityTypeConfiguration<Teacher>
    {
        private const string TableName = "Teachers";
        public void Configure(EntityTypeBuilder<Teacher> builder)
        {
            //Первичный ключ
            builder
                .HasKey(p => p.Id)
                .HasName($"pl_{TableName}_teacher_id");

            //Автоинкрементация
            builder.Property(p => p.Id)
                .ValueGeneratedOnAdd()
                .HasColumnName("Teacher_Id")
                .HasComment("Id преподавателя");

            // Колонка имени
            builder.Property(p => p.FirstName)
                .IsRequired()
                .HasColumnName("First_Name")
                .HasColumnType(ColumnType.String).HasMaxLength(20)
                .HasComment("Имя преподавателя");

            // Колонка отчества
            builder.Property(p => p.LastName)
                .IsRequired()
                .HasColumnName("Last_Name")
                .HasColumnType(ColumnType.String).HasMaxLength(20)
                .HasComment("Фамилия преподавателя");

            // Колонка с уч. степенью
            builder.Property(p => p.DegreeId)
                .IsRequired()
                .HasColumnName("Degree_Id")
                .HasColumnType(ColumnType.Int)
                .HasComment("Id ученой степени");

            // Настройка связи (при удалении летит в null)
            builder.HasOne(p => p.AcademicDegree)
                .WithMany()
                .HasForeignKey(p => p.DegreeId)
                .HasConstraintName("fk_f_degree_id")
                .OnDelete(DeleteBehavior.Cascade); //Пересмотреть

            builder.ToTable(TableName)
                .HasIndex(p => p.DegreeId, $"idx_{TableName}_fk_f_degree_id");

            builder.Navigation(p => p.AcademicDegree)
                .AutoInclude();

            // Колонка с должностью
            builder.Property(p => p.PositionId)
                .IsRequired()
                .HasColumnName("Position_Id")
                .HasColumnType(ColumnType.Int)
                .HasComment("Id занимаемой должности");

            // Настройка связи (при удалении препод тоже удаляется)
            builder.HasOne(p => p.Position)
                .WithMany()
                .HasForeignKey(p => p.PositionId)
                .HasConstraintName("fk_f_position_id")
                .OnDelete(DeleteBehavior.Cascade);

            builder.ToTable(TableName)
                .HasIndex(p => p.DegreeId, $"idx_{TableName}_fk_f_position_id");

            builder.Navigation(p => p.Position)
                .AutoInclude();

            //  Колонка с кафедрой
            builder.Property(p => p.DepartmentId)
                .IsRequired()
                .HasColumnName("Department_Id")
                .HasColumnType(ColumnType.Int)
                .HasComment("Id кафедры");

            // Настройка связи (при удалении летит в null)
            builder.HasOne(p => p.Department)
                .WithMany()
                .HasForeignKey(p => p.DepartmentId)
                .HasConstraintName("fk_f_department_id")
                .OnDelete(DeleteBehavior.Cascade); //Пересмотреть

            builder.ToTable(TableName)
                .HasIndex(p => p.DepartmentId, $"idx_{TableName}_fk_f_department_id");
        }

    }
}
